### ISSUE TITLE
Refactor: Tests

### ISSUE THREAD

**Issue by inancgumus** [grafana/xk6-browser#140]
I'm working on browser test refactoring. Hopefully, this will allow us to reason about the tests better and write them quickly, make them run fast, and provide a step for reducing circular dependencies.

**Update (2021-12-14):** Bugs surfaced after this refactoring, especially when running tests in parallel. See PR #141 for details.

## Planned Improvements:

~✅ 1. Make all eligible tests non-subtests and parallel.~

~According to my experiments, there is a 20-40% performance increase even though we opened up a new browser in each test (instead of once for all subtests).~

There are some high-concurrency bugs discovered when running tests in parallel (see PR #141 for details). So we decided to disable parallel test runs until we fix these bugs.

✅ 2. Prevent stuttering and move the test browser creation to the `tests` package (_where it belongs_).

```golang
// Before:
bt := browsertest.NewBrowserTest(t)

// After:
b := newTestBrowser(t)
```

✅ 3. Automatically close the browser when a test ends. So we don't have to do error-pronish `Cleanup` calls.

```golang
// Before:
bt := browsertest.NewBrowserTest(t)
t.Cleanup(bt.Close)

// After:
// The test browser get automatically closed.
```

✅ 4. Provide a `withFileServer()` method that automatically starts a file server for static files.

```golang
// Before:
//go:embed html/empty.html
var HTMLEmptyPage string

//go:embed html/wait_until.html
var HTMLWaitUntilPage string

func setupHandlersForHTMLFiles(bt *browsertest.BrowserTest) {
	...
}
...
func TestX(..) {
	bt := browsertest.NewBrowserTest(t)
	..
	setupHandlersForHTMLFiles()
	..
	p.Goto(bt.HTTPMultiBin.ServerHTTP.URL+"/wait_until.html", ...)
	...
}

// After:
func TestX(..) {
	b := newTestBrowser(t, withFileServer())

	// A URL method is provided as well for easy accessing the test HTTP server
	p.Goto(b.URL("/static/wait_until.html"), ...)
 	...

	// We can further simplify this with:
	p.Goto(b.staticURL("wait_until.html"), ...)
}
```

✅ 5. Provide a `withHandler` method. So we can attach new HTTP handlers to the HTTP test server.

```golang
// Before
bt.HTTPMultiBin.Mux.Handle("/favicon.ico", ...)

// After:
// after enabling the HTTP test server
b.withHandler("/favicon.ico", ...)
```

✅ 6. Make `TestBrowser` launch options configurable.

```golang
// Before:
newTestBrowser(t)

// After:
newTestBrowser(t, withLaunchOptions{
    SlowMo: "500ms",
    Timeout: "30s",
})

// Or:
opts := defaultLaunchOpts()
opts.SlowMo = "500ms"
newTestBrowser(t, opts)
```

✅ 7. Do not initialize the test HTTP server by default and speed up tests more (_test now finish 15% faster_).

```golang
tb := newTestBrowser(t, withHTTPServer())
```

✅ 8. Ability to set the context from a test

```golang
tb := newTestBrowser(t, withContext(context.Background()))
```

~~This is not entirely possible because k6 `HTTPMultiBin` creates its own context! 🤷~~

✅  9. Remove page cleanups because they are not needed for _browser per test_ tests.

✅  10. Unexport types and methods.

```golang
// Before:
NewTestBrowser(t)

// After:
newTestBrowser(t)
```
